File: wfp-spider/src/main/java/net/hoyoung/wfp/spider/comweb/process/Sha1Distinct.java
package net.hoyoung.wfp.spider.comweb.process;

import java.util.List;
import java.util.Set;

import org.apache.commons.collections.CollectionUtils;
import org.bson.Document;
import org.bson.conversions.Bson;
import org.bson.types.ObjectId;

import com.google.common.collect.Lists;
import com.google.common.collect.Sets;
import com.mongodb.client.MongoCollection;
import com.mongodb.client.MongoCursor;
import com.mongodb.client.model.Filters;
import com.mongodb.client.model.Projections;

import net.hoyoung.wfp.core.utils.MongoUtil;
import net.hoyoung.wfp.spider.comweb.ComWebConstant;
import net.hoyoung.wfp.spider.comweb.bo.ComPage;

public class Sha1Distinct {

	public static void main(String[] args) {
		// String dbName = ComWebConstant.DB_NAME + "_test";
		String dbName = ComWebConstant.DB_NAME;

		List<String> list = MongoUtil.getCollectionNames(dbName, "\\d{6}");
		if (CollectionUtils.isEmpty(list)) {
			return;
		}
		int cnt = 0;
		for (String collectionName : list) {
			if (++cnt > 450)
				break;
			MongoCollection<Document> collection = MongoUtil.getCollection(dbName, collectionName);
			Bson filters = Filters.and(Filters.exists(ComPage.CONTENT_SHA1));
			long total = collection.count(filters);
			System.out.println(String.format("%d\tprocess collection %s,total=%d", cnt, collectionName, total));
			MongoCursor<Document> iterator = collection.find(filters)
					.projection(Projections.include(ComPage.CONTENT_SHA1)).iterator();
			Set<String> sha1Set = Sets.newHashSet();
			List<ObjectId> oids = Lists.newArrayList();
			try {
				while (iterator.hasNext()) {
					Document document = iterator.next();
					String sha1 = document.getString(ComPage.CONTENT_SHA1);
					if (sha1Set.contains(sha1)) {
						oids.add(document.getObjectId("_id"));
					} else {
						sha1Set.add(sha1);
					}
				}
			} finally {
				iterator.close();
			}
			System.out.println(String.format("collection %s delete %d", collectionName, oids.size()));
			if (CollectionUtils.isNotEmpty(oids)) {
				collection.deleteMany(Filters.in("_id", oids));
			}
			// break;
		}
	}

}

File: wfp-python/compage/mongo/sha1_distinct.py
from common.mongo import mongo_cli
import re
import pymongo

if __name__ == '__main__':
    db = mongo_cli.get_database('wfp_com_page')
    sha1_distinct_collection = mongo_cli.get_database('tmp').get_collection('sha1_distinct')
    sha1_distinct_collection.create_index([('stockCode', pymongo.ASCENDING)], unique=True)
    collection_names = [c for c in db.collection_names(include_system_collections=False) if re.match('^\d{6}$', c)]
    cnt = 0
    for collection_name in collection_names:
        cnt += 1
        if sha1_distinct_collection.count({'stockCode': collection_name}) > 0:
            print('\r%d-%s\tfinished' % (cnt, collection_name), end='')
            print()
            continue
        collection = db.get_collection(collection_name)
        total = collection.count()
        filters = {
            'contentSha1': {
                '$exists': True
            }
        }
        total_html = collection.count(filters)
        del_list = []
        sha1_set = set()
        cnt2 = 0
        for doc in collection.find(filters, projection={'contentSha1': 1}):
            cnt2 += 1
            contentSha1 = doc['contentSha1']
            print('\r%d-%s\t%d/%d\t%d' % (cnt, collection_name, cnt2, total_html, len(del_list)), end='')
            if contentSha1 in sha1_set:
                del_list.append(doc['_id'])
            else:
                sha1_set.add(contentSha1)
        if len(del_list) > 0:
            collection.delete_many({
                '_id': {
                    '$in': del_list
                }
            })
        sha1_distinct_collection.insert_one(
            {'stockCode': collection_name, 'total': total, 'total_html': total_html, 'total_del': len(del_list)})
        print()
        # break
    mongo_cli.close()
